### PR TITLE
feat/hiervis

### PR DESCRIPTION
**Pull request by kartoFlane**
Cleaned up version of previous pull request (#1)

...Forgot to replace temp name of the last commit with something meaningful. Extended description contains more information on what was actually changed.

### CODE REVIEW

File: src/basic_hierarchy/interfaces/Instance.java [hunk 1]
```diff
@@ -1,9 +1,30 @@
 package basic_hierarchy.interfaces;
 
-public interface Instance {
+/**
+ * An instance is a singular entity that can be placed in the broader context of a {@link Hierarchy}.
```
Comment (lpolech) [inline src/basic_hierarchy/interfaces/Instance.java]: My proposition:
An instance is a singular entity (a single observation/data point) that can be placed in the {@link Node} that are placed in the broader context in a {@link Hierarchy}.

File: src/basic_hierarchy/interfaces/Hierarchy.java [hunk 2]
```diff
@@ -1,12 +1,54 @@
...
 	public int getNumberOfClasses();
-	public int getClassCount(String className, boolean withClassHierarchy);
+
+	/**
+	 * Returns the total number of instances in this hierarchy, ie. sum of instances in all groups.
+	 */
 	public int getNumberOfInstances();
+
+	/**
+	 * Returns the number of groups within the specified class.
```
Comment (lpolech) [inline src/basic_hierarchy/interfaces/Hierarchy.java]: Returns the number of instances that are assigned to {@code className} class within the whole hierarchy.

File: src/basic_hierarchy/interfaces/Instance.java [hunk 3]
```diff
@@ -1,9 +1,30 @@
...
+public interface Instance
+{
+	/**
+	 * @return the name of this particular instance
+	 */
 	public String getInstanceName();
-	public String getNodeId();
+
+	/**
+	 * @return data columns associated with this instance
```
Comment (lpolech) [inline src/basic_hierarchy/interfaces/Instance.java]: @return data row (feature values) associated with this instance

File: src/basic_hierarchy/interfaces/Hierarchy.java [hunk 4]
```diff
@@ -1,12 +1,54 @@
...
 	public String[] getClasses();
+
+	/**
+	 * Returns the number of ground-truth classes. Equivalent to {@code getClasses().length}.
+	 */
 	public int getNumberOfClasses();
-	public int getClassCount(String className, boolean withClassHierarchy);
+
+	/**
+	 * Returns the total number of instances in this hierarchy, ie. sum of instances in all groups.
```
Comment (lpolech) [inline src/basic_hierarchy/interfaces/Hierarchy.java]: @return the total number of instances in this hierarchy, ie. sum of instances in all groups.

File: src/basic_hierarchy/interfaces/Hierarchy.java [hunk 5]
```diff
@@ -1,12 +1,54 @@
...
+	 */
 	public int getNumberOfGroups();
+
+	/**
+	 * Returns an array of classes present in this hierarchy.
+	 */
 	public String[] getClasses();
+
+	/**
+	 * Returns the number of ground-truth classes. Equivalent to {@code getClasses().length}.
```
Comment (lpolech) [inline src/basic_hierarchy/interfaces/Hierarchy.java]: @return the number of ground-truth classes. Equivalent to {@code getClasses().length}.

File: src/basic_hierarchy/interfaces/Hierarchy.java [hunk 6]
```diff
@@ -1,12 +1,54 @@
...
+	 */
 	public Node[] getGroups();
+
+	/**
+	 * Returns the number of groups in this hierarchy. Equivalent to {@code getGroups().length}.
+	 */
 	public int getNumberOfGroups();
+
+	/**
+	 * Returns an array of classes present in this hierarchy.
```
Comment (lpolech) [inline src/basic_hierarchy/interfaces/Hierarchy.java]: use @return

File: src/basic_hierarchy/interfaces/Hierarchy.java [hunk 7]
```diff
@@ -1,12 +1,54 @@
...
+	 */
 	public Node getRoot();
+
+	/**
+	 * Returns an array of groups of instances.
+	 */
 	public Node[] getGroups();
+
+	/**
+	 * Returns the number of groups in this hierarchy. Equivalent to {@code getGroups().length}.
```
Comment (lpolech) [inline src/basic_hierarchy/interfaces/Hierarchy.java]: use @return

File: src/basic_hierarchy/interfaces/Hierarchy.java [hunk 8]
```diff
@@ -1,12 +1,54 @@
...
+ */
+public interface Hierarchy
+{
+	/**
+	 * Returns the root node of this hierarchy.
+	 */
 	public Node getRoot();
+
+	/**
+	 * Returns an array of groups of instances.
```
Comment (lpolech) [inline src/basic_hierarchy/interfaces/Hierarchy.java]: use @return

File: src/basic_hierarchy/interfaces/Hierarchy.java [hunk 9]
```diff
@@ -1,12 +1,54 @@
 package basic_hierarchy.interfaces;
 
-public interface Hierarchy {
+/**
+ * A hierarchy is a collection of {@link Node}s, all of which share one common ancestor (the root group).
+ */
+public interface Hierarchy
+{
+	/**
+	 * Returns the root node of this hierarchy.
```
Comment (lpolech) [inline src/basic_hierarchy/interfaces/Hierarchy.java]: use @return

File: src/basic_hierarchy/interfaces/Instance.java [hunk 10]
```diff
@@ -1,9 +1,30 @@
...
+	 */
 	public double[] getData();
+
+	/**
+	 * @return id of the true class (group) to which this instance has been assigned ('ground truth').
```
Comment (lpolech) [inline src/basic_hierarchy/interfaces/Instance.java]: @return id of the true class (group) from which this instance comes from ('ground truth').

File: src/basic_hierarchy/interfaces/Instance.java [hunk 11]
```diff
@@ -1,9 +1,30 @@
...
+	 */
 	public String getTrueClass();
-	public void setNodeId(String id);
+
+	/**
+	 * @return id of the group to which this instance has been assigned ('assigned class').
+	 */
+	public String getNodeId();
+
+	public void setNodeId( String id );
```
Comment (lpolech) [inline src/basic_hierarchy/interfaces/Instance.java:36]: missing javadoc?

File: src/basic_hierarchy/interfaces/Node.java [hunk 12]
```diff
@@ -2,36 +2,94 @@
 
 import java.util.LinkedList;
 
-public interface Node {	
+
+/**
+ * A group is a collection of {@link Instance}s which belong to the same class.
```
Comment (lpolech) [inline src/basic_hierarchy/interfaces/Node.java]: A {@link Node} is a collection of {@link Instance}s which were assign to one {@link Node} by i.e. some method or generator.

File: src/basic_hierarchy/interfaces/Node.java [hunk 13]
```diff
@@ -2,36 +2,94 @@
...
+ */
+public interface Node
+{
+	/**
+	 * @return the id of this group.
```
Comment (lpolech) [inline src/basic_hierarchy/interfaces/Node.java]: @return the id of this node.

File: src/basic_hierarchy/interfaces/Node.java [hunk 14]
```diff
@@ -2,36 +2,94 @@
```
Comment (lpolech) [inline src/basic_hierarchy/interfaces/Node.java:2]: In this file you should replace "group" with "node" in the docs

File: src/basic_hierarchy/interfaces/Node.java [hunk 15]
```diff
@@ -2,36 +2,94 @@
...
+	/**
+	 * @return the child groups of this group.
+	 */
 	public LinkedList<Node> getChildren();
-	
-	/*
-	 * Should returns instances belonging to this particular node
+
+	/**
+	 * Returns instances which belong to this particular group.
```
Comment (lpolech) [inline src/basic_hierarchy/interfaces/Node.java]: use @return

File: src/basic_hierarchy/interfaces/Node.java [hunk 16]
```diff
@@ -2,36 +2,94 @@
...
 	 */
 	public LinkedList<Instance> getNodeInstances();
-	/*
-	 * Should returns instances belonging to this node and its predecessors 
+
+	/**
+	 * Returns instances which belong to this group or its subgroups.
```
Comment (lpolech) [inline src/basic_hierarchy/interfaces/Node.java]: use @return

File: src/basic_hierarchy/interfaces/Node.java [hunk 17]
```diff
@@ -2,36 +2,94 @@
...
 	 */
 	public LinkedList<Instance> getSubtreeInstances();
-	/*
-	 * Could be node centroid or medoid
+
+	/**
+	 * Returns a node centroid or medoid.
```
Comment (lpolech) [inline src/basic_hierarchy/interfaces/Node.java]: @return node representation e.g. its centroid or medoid.

File: src/basic_hierarchy/interfaces/Node.java [hunk 18]
```diff
@@ -2,36 +2,94 @@
...
-	public void setRepresentation(Instance representation);
+	/**
+	 * Sets the representation instance of this node.
+	 * 
+	 * @param representation
+	 *            representation instance
+	 */
+	public void setRepresentation( Instance representation );
 
 	public void printSubtree();
```
Comment (lpolech) [inline src/basic_hierarchy/interfaces/Node.java:98]: javadoc?

**Review by lpolech — CHANGES_REQUESTED** [review]
Find detailed comments in files

**Review by lpolech — APPROVED** [review]
I like that implementation!

**Review by lpolech — CHANGES_REQUESTED** [review]
Why did you change nodeId to assignedClass? What is the purpose of this change? If so maybe it is worth to change class property from nodeId to assignedClass?

File: src/basic_hierarchy/implementation/BasicNode.java [hunk 19]
```diff
@@ -2,114 +2,135 @@
...
 	}
 
 	@Override
-	public LinkedList<Instance> getSubtreeInstances() {
-		LinkedList<Instance> subtreeInstances = new LinkedList<Instance>(instances);
-		
-		for(Node child: children)
-		{
-			subtreeInstances.addAll(child.getSubtreeInstances());
+	public LinkedList<Instance> getSubtreeInstances()
```
Comment (lpolech) [inline src/basic_hierarchy/implementation/BasicNode.java:120]: What is the purpose of this change?

File: src/basic_hierarchy/implementation/BasicNode.java [hunk 20]
```diff
@@ -2,114 +2,135 @@
...
+	{
 		this.instances = instances;
 	}
 
 	@Override
-	public void setRepresentation(Instance representation)
-    {
-        this.representation = representation;
-    }
+	public void setRepresentation( Instance representation )
```
Comment (lpolech) [inline src/basic_hierarchy/implementation/BasicNode.java:84]: I would prefer expand tab with 4 spaces

**Review by lpolech — CHANGES_REQUESTED** [review]

**Review by lpolech — APPROVED** [review]

File: src/basic_hierarchy/implementation/BasicHierarchy.java [hunk 21]
```diff
@@ -1,101 +1,183 @@
...
-		int arrayIndex = 0;
-		for(String key: sortedKeyes)
-		{
-			classes[arrayIndex] = key;
-			classesCount[arrayIndex] = eachClassWithCount.get(key);
-			arrayIndex++;
+		classCounts = new int[eachClassWithCount.size()];
+
+		LinkedList<String> sortedKeys = new LinkedList<String>( eachClassWithCount.keySet() );
+		sortedKeys.sort( new StringIdComparator() );
```
Comment (lpolech) [inline src/basic_hierarchy/implementation/BasicHierarchy.java]: Unfortunately this code needs to support java 1.7.

**Review by lpolech — CHANGES_REQUESTED** [review]

File: src/basic_hierarchy/reader/GeneratedCSVReader.java [hunk 22]
```diff
@@ -1,152 +1,195 @@
...
-	 * Assume that:
-	 * "The first node listed is always
-     * the root node, it is also the only node without a parent and the node specified
-     * in the first section to be the root of the tree. The nodes are given in depth first
-     * order."
-	 * @see interfaces.DataReader#load(java.lang.String)
+public class GeneratedCSVReader implements DataReader
+{
+	/**
+	 * This method assumes that data are generated using Michał Spytkowski's data generator, using TSSB method.
```
Comment (lpolech) [inline src/basic_hierarchy/reader/GeneratedCSVReader.java]: What about adding a reference to the article that is describing this generator? https://arxiv.org/abs/1606.05681

File: src/basic_hierarchy/reader/GeneratedCSVReader.java [hunk 23]
```diff
@@ -1,152 +1,195 @@
...
-public class GeneratedCSVReader implements DataReader {
-	/*
-	 * I assume that data are generated using Michal Spytkowski's data generator with the use of TSSB method.
-	 * Assume that:
-	 * "The first node listed is always
-     * the root node, it is also the only node without a parent and the node specified
-     * in the first section to be the root of the tree. The nodes are given in depth first
-     * order."
-	 * @see interfaces.DataReader#load(java.lang.String)
+public class GeneratedCSVReader implements DataReader
```
Comment (lpolech) [inline src/basic_hierarchy/reader/GeneratedCSVReader.java:27]: What about splitting load function into smaller functions?

File: src/basic_hierarchy/reader/GeneratedARFFReader.java [hunk 24]
```diff
@@ -19,8 +21,14 @@
 public class GeneratedARFFReader implements DataReader {
 
 	@Override
-	public Hierarchy load(String filePath, boolean withInstancesNameAttribute, boolean withClassAttribute,
-						  boolean fillBreathGaps, boolean useSubtree) {
+	public Hierarchy load(
```
Comment (lpolech) [inline src/basic_hierarchy/reader/GeneratedARFFReader.java:24]: What about splitting load function into smaller functions? And maybe find some common abstractions to be used both in CSV and ARFF readers

File: src/basic_hierarchy/common/HierarchyBuilder.java [hunk 25]
```diff
@@ -0,0 +1,395 @@
...
+
+		if ( fixBreadthGaps ) {
+			groups = fixBreadthGaps( root, groups, useSubtree );
+		}
+
+		for ( BasicNode g : groups ) {
+			g.recalculateCentroid( useSubtree );
+		}
+
+		groups.sort( new NodeIdComparator() );
```
Comment (lpolech) [inline src/basic_hierarchy/common/HierarchyBuilder.java]: We need java 1.7 support ;(

File: src/basic_hierarchy/common/HierarchyBuilder.java [hunk 26]
```diff
@@ -0,0 +1,395 @@
...
+	 * Checks whether the two groups are directly related (parent-child).
+	 * This method returns false if both ids point to the same group.
+	 * 
+	 * @param parentIds
+	 *            ID segments of the group acting as parent
+	 * @param childIds
+	 *            ID segments of the group as child
+	 * @return whether the two groups are in fact in a direct parent-child relationship.
+	 */
+	private static boolean areGroupsDirectlyRelated( String[] parentIds, String[] childIds )
```
Comment (lpolech) [inline src/basic_hierarchy/common/HierarchyBuilder.java]: I would suggest to change the name of this method to relate that we are comparing strings and checking if they are showing a parent->child relation

I don't like, when two methods have the same names with different parameters, because it is the sign that these two methods are doing different things.

File: src/basic_hierarchy/common/HierarchyBuilder.java [hunk 27]
```diff
@@ -0,0 +1,395 @@
...
+	 * 
+	 * @param root
+	 *            the root group
+	 * @param groups
+	 *            the original collection of groups
+	 * @param useSubtree
+	 *            whether the cntroid calculation should also include child groups' instances.
+	 * @return the complete 'fixed' collection of groups, filled with artificial groups
+	 */
+	private static List<BasicNode> fixBreadthGaps( BasicNode root, List<BasicNode> groups, boolean useSubtree )
```
Comment (lpolech) [inline src/basic_hierarchy/common/HierarchyBuilder.java]: Can you split this method into smaller ones?

File: src/basic_hierarchy/common/HierarchyBuilder.java [hunk 28]
```diff
@@ -0,0 +1,395 @@
...
+	 * @param root
+	 *            the root group
+	 * @param groups
+	 *            the original collection of groups
+	 * @param useSubtree
+	 *            whether the cntroid calculation should also include child groups' instances.
+	 * @return the complete 'fixed' collection of groups, filled with artificial groups
+	 */
+	private static List<BasicNode> fixDepthGaps( BasicNode root, List<BasicNode> groups, boolean useSubtree )
+	{
```
Comment (lpolech) [inline src/basic_hierarchy/common/HierarchyBuilder.java:129]: What about splitting this method into smaller ones?

File: src/basic_hierarchy/common/HierarchyBuilder.java [hunk 29]
```diff
@@ -0,0 +1,395 @@
...
+		pendingGroups.add( root );
+
+		while ( !pendingGroups.isEmpty() ) {
+			BasicNode currentGroup = pendingGroups.remove();
+
+			List<Node> children = currentGroup.getChildren();
+			List<Node> newChildren = new LinkedList<Node>();
+
+			for ( int i = 0; i < children.size(); ++i ) {
+				Node childGroup = children.get( i );
```
Comment (lpolech) [inline src/basic_hierarchy/common/HierarchyBuilder.java]: Where do we assure that children will have null values when there will be no child node?

**Review by lpolech — CHANGES_REQUESTED** [review]

File: src/basic_hierarchy/common/HierarchyBuilder.java [hunk 30]
```diff
@@ -0,0 +1,395 @@
...
+	 * for missing IDs.
+	 * 
+	 * @param root
+	 *            the root group
+	 * @param groups
+	 *            the original collection of groups
+	 * @param fixBreadthGaps
+	 *            whether the hierarchy fixing algorithm should also fix gaps in breadth, not just depth.
+	 * @param useSubtree
+	 *            whether the centroid calculation should also include child groups' instances.
```
Comment (lpolech) [inline src/basic_hierarchy/common/HierarchyBuilder.java]: Add:
when set to {@code true} then all the objects from subnodes are also regarded as belonging to its supernodes.

**Review by lpolech — CHANGES_REQUESTED** [review]

File: src/basic_hierarchy/implementation/BasicNode.java (hunk 19 above)
Comment (kartoFlane) [reply src/basic_hierarchy/implementation/BasicNode.java:120]: This is a simple variable rename, combined with code formatting changes made by autoformatter. Must've slipped through the purge of cosmetic changes.

File: src/basic_hierarchy/reader/GeneratedCSVReader.java (hunk 23 above)
Comment (kartoFlane) [reply src/basic_hierarchy/reader/GeneratedCSVReader.java:27]: Should be possible. Will look into this later.

File: src/basic_hierarchy/reader/GeneratedCSVReader.java (hunk 22 above)
Comment (kartoFlane) [reply src/basic_hierarchy/reader/GeneratedCSVReader.java]: Done.

File: src/basic_hierarchy/common/HierarchyBuilder.java (hunk 29 above)
Comment (kartoFlane) [reply src/basic_hierarchy/common/HierarchyBuilder.java]: Now that I look at this, I feel like I might've made an incorrect assumption here. I'll check this later.

File: src/basic_hierarchy/implementation/BasicNode.java (hunk 19 above)
Comment (lpolech) [reply src/basic_hierarchy/implementation/BasicNode.java:120]: Don't you like the name "subtreeInstances"?

Comment (kartoFlane) [reply src/basic_hierarchy/implementation/BasicNode.java:120]: I usually prefer to name the variable holding the return value as 'result', at least in simple functions.

File: src/basic_hierarchy/common/HierarchyBuilder.java (hunk 29 above)
Comment (kartoFlane) [reply src/basic_hierarchy/common/HierarchyBuilder.java]: Indeed, this method was not working correctly. Fixed in ea3ad3b.

File: src/basic_hierarchy/common/HierarchyBuilder.java (hunk 26 above)
Comment (kartoFlane) [reply src/basic_hierarchy/common/HierarchyBuilder.java]: Done. 57cd617

File: src/basic_hierarchy/common/HierarchyBuilder.java (hunk 28 above)
Comment (kartoFlane) [reply src/basic_hierarchy/common/HierarchyBuilder.java:129]: Done. 5a3763d

File: src/basic_hierarchy/common/HierarchyBuilder.java (hunk 27 above)
Comment (kartoFlane) [reply src/basic_hierarchy/common/HierarchyBuilder.java]: Done. 87b4dc5

File: src/basic_hierarchy/common/HierarchyBuilder.java (hunk 25 above)
Comment (kartoFlane) [reply src/basic_hierarchy/common/HierarchyBuilder.java]: Done. cb25b20
